File: brainfuck.c
#define NELEMS(x) (sizeof(x) / sizeof((x)[0]))

#ifndef VERBOSE
#define VERBOSE 0
#endif
#ifndef S
#define S 500000
#endif
#ifndef C
#define C 500000
#endif
#ifndef STACKWIDTH
#define STACKWIDTH 15
#endif


typedef enum
{
    JUMP,
    LOOP,
    CNT,
    HALT
} signal;

#define JUMP 1
#define LOOP 2
#define CNT 3
#define HALT 4

#include <stdio.h>
#include <string.h>


void printStack(char* stack, const size_t n)
{
    int i = 0;
    printf("\n ");
    for (; i < n-1; ++i)
        printf("%3i, ", stack[i]);
    printf("%3i\n", stack[i+1]);
}

void dispState(char* stack, const size_t n, int pointer) {
    printf("\nStack state:\n");
    printStack(stack, STACKWIDTH);
    // char str_pointer[STACKWIDTH*5];
    // memset(str_pointer, ' ', NELEMS(str_pointer));
    // str_pointer[(pointer*4)] = '^';
    printf("%d\n", pointer);
}

unsigned int matchLoop(char* code, unsigned int inst) {
    unsigned int i = inst-1;
    // if (VERBOSE) printf("%d\n", i);
    int nested = 1;
    while (i > 0)
    {
        // printf("%d\n", i);
        if (nested == 0)
        {
            return i+1;
        }
        if (code[i] == '[')
        {
            nested-=1;
        }
        if (code[i] == ']')
        {
            nested+=1;
        }
        i--;
    }
    return 0;
}

unsigned int matchJump(char* code, unsigned int inst) {
    unsigned int i = inst+1;
    int nested = 1;
    while (code[i])
    {
        // if (VERBOSE) printf("%d\n", i);
        if (nested == 0)
        {
            return i;
        }
        if (code[i] == ']')
        {
            nested-=1;
        }
        if (code[i] == '[')
        {
            nested+=1;
        }
        i++;
    }
    return 0;
}

signal runInstruction(char* stack, unsigned int * pointer, char inst)
{
    switch (inst)
    {
    case '+':
        stack[*pointer]++;
        return CNT;
        break;
    case '-':
        stack[*pointer]--;
        return CNT;
        break;
    case '>':
        (*pointer)++;
        return CNT;
        break;
    case '<':
        (*pointer)--;
        return CNT;
        break;
    case '[':
        if (VERBOSE) printf("%d",stack[ *pointer]);
        if (stack[*pointer] == 0) return JUMP;
        else return CNT;
        break;
    case ']':
        return LOOP;
        break;
    case '.':
        if (VERBOSE) printf("printing ");
        printf("%c", stack[*pointer]);
        if (VERBOSE) printf("\n");
        return CNT;
        break;
    case ',':
        scanf("%c", &stack[*pointer]);
        return CNT;
        break;
    default:
        return HALT;
        break;
    }
}

void getCode(FILE *file, char code[], int *code_len)
{

    int c;
    int i = 0;
    while ((c = fgetc(file)) != EOF)
    {
        if (strchr("<>+-[],.", c))
        {
            code[i] = c;
            i++;
        }
    }
    *code_len = i;
    fclose(file);
}


int main(int argc, char const *argv[])
{
    FILE *file;
    if (argc < 2) {
        printf("No input file specified, reading from stdin\n");
        file = stdin;
    }
    else
    {
        file = fopen(argv[1], "r");
        if (file==NULL)
        {
            printf("File %s not found\n", argv[1]);
            return 1;
        }
    }


    // Create a str with the code
    char code[C];
    memset(code, '\0', NELEMS(code));
    int lastInst;
    getCode(file, code, &lastInst);

    printf("Executing: \n%s\n", code);
    char stack[S];
    memset(stack, 0, NELEMS(stack));
    unsigned int stackPointer = 0;
    unsigned int insPointer = 0;

    while (insPointer < lastInst)
    {
        if (VERBOSE) printf("%x, %d, %c\n", stackPointer, insPointer, code[insPointer]);
        if (VERBOSE) printStack(stack, NELEMS(stack));
        // int acc = runInstruction(stack, &stackPointer, code[insPointer]);
        switch (runInstruction(stack, &stackPointer, code[insPointer]))
        {
        case CNT:
            insPointer++;
            break;
        case JUMP:
            if (VERBOSE) printf("matching jump: ");
            insPointer = matchJump(code, insPointer);
            if (VERBOSE) printf("%d\n", insPointer);
            break;
        case LOOP:
            if (VERBOSE) printf("matching loop: ");
            insPointer = matchLoop(code, insPointer);
            if (VERBOSE) printf("%d\n", insPointer);
            break;
        case HALT:
            insPointer = 1+lastInst;
            break;
        default:
            break;
        }
    }

    dispState(stack, NELEMS(stack), stackPointer);

    return 0;
}

File: brainfuck.py
#! /usr/bin/python3

# Copyright (C) 2019  Alberto Di Biase

import sys

JUMP = 1
CNT = 2
LOOP = 3

class BrainFuckInter():
    def __init__(self, stack_size = 0):
        self.stack_size = stack_size
        self.pointer = 0
        if stack_size == 0:
            self.stack_inf = True
            self.stack = [0]
        else:
            self.stack = [0 for _ in range(stack_size)]
            self.stack_inf = False

        self.inst = {
            '+' : self.inc,
            '-' : self.dec,
            '>' : self.mvr,
            '<' : self.mvl,
            '.' : self.show,
            ',' : self.scan,
            '[' : self.loop_start,
            ']' : self.loop_end
        }

    def resize(self):
        # print(self.pointer, len(self.stack), self.pointer - len(self.stack))
        if self.pointer > len(self.stack)-1:
            self.stack += [0 for _ in range(self.pointer - len(self.stack) + 2)]

    def inc(self):
        if self.stack_inf: self.resize()
        self.stack[self.pointer] += 1
        return CNT

    def dec(self):
        if self.stack_inf: self.resize()
        self.stack[self.pointer] -= 1
        return CNT

    def show(self):
        print("{}".format(chr(self.stack[self.pointer])), end="")
        return CNT

    def scan(self):
        c = input()
        self.stack[self.pointer] = c
        return CNT

    def mvr(self):
        self.pointer+=1
        if self.stack_inf: self.resize()
        return CNT

    def mvl(self):
        self.pointer-=1
        if self.stack_inf: self.resize()
        return CNT

    def loop_start(self):
        if self.stack[self.pointer] == 0:
            return JUMP
        else:
            return CNT

    def loop_end(self):
        return LOOP

    def decode_and_run(self, inst):
        return self.inst[inst]()

    def __repr__(self):
        stack = ""
        for s in self.stack:
            stack += "{:3d}, ".format(s)
        stack = '\n' + str(self.stack) + '\n' \
            + ' ' + "   "*self.pointer  + "^\n" \
            + "Pointer at: {}".format(self.pointer)
        return stack



def getCode(file):
    code = []
    for line in file:
        for c in line:
            if c in "+-<>.,[]":
                code.append(c)
    return code

def matchJump(code, pointer):
    nested = 1
    i = pointer + 1
    while i < len(code):
        if nested == 0:
            return i
        if code[i] == '[':
            nested += 1
        if code[i] == ']':
            nested -= 1
        i += 1
    return i

def matchLoop(code, pointer):
    nested = 1
    i = pointer - 1
    while i > 0:
        if nested == 0:
            return i+1
        if code[i] == '[':
            nested -= 1
        if code[i] == ']':
            nested += 1
        i -= 1
    return 0


if __name__ == "__main__":
    if len(sys.argv) > 1:
        fp = open(sys.argv[1], 'r')
    else:
        fp = sys.stdin
    debug = False
    if "-d" in sys.argv:
        debug = True
    S = 0
    if "-s" in sys.argv:
        S = int(sys.argv[sys.argv.index("-s")+1])

    code = getCode(fp)
    fp.close()
    s = ""
    for c in code:
        s+=c
    print("Executing:")
    print(s)

    bf = BrainFuckInter(stack_size=S)
    i = 0
    while i < len(code):
        if debug: print(code[i], i)
        if debug: print(bf)
        rst = bf.decode_and_run(code[i])
        if rst == JUMP:
            i = matchJump(code, i)
            if debug: print("Match Jump: ", i)
        elif rst == LOOP:
            i = matchLoop(code, i)
            if debug: print("Match Loop: ", i)
        elif rst == CNT:
            i += 1

    print(bf)